include/swift/SIL/SILConstants.h

```diff
@@ -114,6 +114,12 @@ class SymbolicValue {
 
     /// This represents an index *into* a memory object.
     RK_DerivedAddress,
+
+    /// This represents an array value.
+    RK_Array,
+
+    /// This represents an address of a memory object containing an array.
+    RK_ArrayAddress,
```

Comment:
@marcrasi Sorry for the delay in the responding to this PR. Based on discussions with @devincoughlin, we have a few questions/comments about this PR. We are trying to understand the motivation for the two representations for an array: `RK_Array` and `RK_ArrayAddress`. Swift's representation of an Array, in principle, consists of `struct` which has a pointer to a backing storage that contains the elements. It seems to us that RK_Array (i.e, ArraySymbolicValue) models the backing storage which stores the elements.  RK_ArrayAddress is modeling the more accurate view of the array as a struct with a pointer to the backing storage namely RK_Array. It seems this distinction between an Array and its backing storage becomes important for modeling `_allocateUninitializedArray` function which returns them separately and initializes the backing storage directly through the pointer.  Our first question is, is our understanding right? 

If that is the case, it seems to us that RK_ArrayAddress is possibly the more accurate view of an array and RK_Array can be seen as  a more optimal representation when the distinction between the array and the backing storage can be safely discarded. Therefore, the name `RK_ArrayAddress` seems a bit misleading as it is not representing an address of an Array, but rather a more accurate view of an Array.  Would it be possible to give RK_Array, RK_ArrayAddress, ArraySymbolicValue another name and also add some documentation? For instance, what do you think about, `SymbolicArrayBuffer` for ArraySymbolicValue, `RK_ArrayBuffer` for RK_Array and `RK_ArrayStruct` for RK_ArrayAddress.   

Ideally, I think we can create a new symbolic value for Array that will manage the differences in the representation. But, I think I can do that in another PR, after this lands.

Reply:
I think `SymbolicArrayStorage` and `RK_ArrayStorage` are better alternative names for `SymbolicArrayBuffer` and `RK_ArrayBuffer` I proposed above. This is inline with `_ContiguousArrayStorage` class used in by struct Array.

Reply:
I like these name suggestions, thanks! Your comments and name suggestions are making it clear to me that these RKs are two fundamentally different things that should actually have different `Kind`s.

Also, I'm pretty sure that the place below where you noticed the collapsing is a bug. (Probably not exposed by anything yet, but likely to become exposed when there are more array features.)

Finally, now that it's clear that RK_ArrayAddress is just a struct containing a pointer, why not make it a RK_Aggregate, like all the normal structs? This should simplify things!

So in summary, the total changes from the current state of this PR are:
* `RK_Array` => `RK_ArrayStorage`
* `ArraySymbolicValue` => `SymbolicArrayStorage`
* `RK_ArrayAddress` => eliminated, replaced by the existing `RK_Aggregate`

I haven't tried this out yet so there may be some difficulty I'm not seeing. So I'll try it out soon and let you know how it goes.

Reply:
Sounds great. :-)

Reply:
On reflection, I think that it may be a good idea to have a special kind for "Array" aggregates (in addition to `ArrayStorage` kind) and not treat them as general aggregates. In fact, I even think we should have special `ArraySymbolicValue` (in addition to `SymbolicArrayStorage`). The following are some reasons for that:

* We don't want to make it appear that the interpreter supports arbitrary aggregates with pointers.  Having a special "kind" for such an aggregate, makes this clear. 

* `get/setIndexedElement` would need to handle `addresses` (which could now appear inside aggregates). Normally, when there are addresses, `setIndexedElement` would be expected to update the underlying memory object in place, as we want everything that aliases with the pointer to also see the changes when `setIndexedElement` completes. But, arrays are COW and provide a value semantics.  This means we have to special case the `address` case for arrays, like creating a new memory object and not in-place updating the underlying memory object. This would appear confusing, as it is only correct for a type like Array that uses pointers underneath but provides a value semantics. 

To address both of this, we may rather have a new symbolic value for "array" which just stores a single `DirectAddressValue` or even just a `SymbolicMemoryObject` containing a `SymbolicArrayStorage`.   This `ArraySymbolicValue` may support operations like `getAddressOfStorage()` which creates a direct address from the underlying SymbolicMemoryObject, and `getStorage()` which returns the storage. `setIndexedElement` can just retrieve and (immutably) update the array storage and create a new ArraySymbolicValue with the new storage and return it.  This design essentially means that the only way to in-place update an `ArraySymbolicValue` instance  is to use `getAddressOfStorage` and update the storage through the address. This is also the *only way* AFAIK in Swift also to mutably update an array storage in place, and is done only to initialize the array.